claat/parser/md/html.go

```diff
@@ -0,0 +1,252 @@
+// Copyright 2016 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package md
+
+import (
+	"bytes"
+	"strings"
+
+	"golang.org/x/net/html"
+	"golang.org/x/net/html/atom"
+)
+
+var (
+	// headerLevel maps HTML tags to their level in parser.HeaderNode.
+	// we -1 as H2 is a new step
+	headerLevel = map[atom.Atom]int{
+		atom.H3: 2,
+		atom.H4: 3,
+		atom.H5: 4,
+		atom.H6: 5,
+	}
+)
+
+// isHeader returns true if hn is one of secondary headers.
+// Step header is not one of them.
+func isHeader(hn *html.Node) bool {
+	_, ok := headerLevel[hn.DataAtom]
+	return ok
+}
+
+func isMeta(hn *html.Node) bool {
+	elem := strings.ToLower(hn.Data)
+	return strings.HasPrefix(elem, metaDuration+metaSep) || strings.HasPrefix(elem, metaEnvironment+metaSep)
+}
+
+func isBold(hn *html.Node) bool {
+	if hn.Type == html.TextNode {
+		hn = hn.Parent
+	}
+	return hn.DataAtom == atom.Strong ||
+		hn.DataAtom == atom.B
+}
+
+func isItalic(hn *html.Node) bool {
+	if hn.Type == html.TextNode {
+		hn = hn.Parent
+	}
+	return hn.DataAtom == atom.Em ||
+		hn.DataAtom == atom.I
+}
+
+func isConsole(hn *html.Node) bool {
+	if hn.Type == html.TextNode {
+		hn = hn.Parent
+	}
+	return hn.DataAtom == atom.Code && hn.Parent.DataAtom != atom.Pre
+}
+
+func isCode(hn *html.Node) bool {
+	if hn.Type == html.TextNode {
+		hn = hn.Parent
+	}
+	return hn.DataAtom == atom.Code && hn.Parent.DataAtom == atom.Pre
+}
+
+func isButton(hn *html.Node) bool {
+	// TODO: implements
+	return false
```

Comment:
Sorry, I missed this the first time.
Do you guys not use the "download" button at all?
It was implemented before in https://github.com/googlecodelabs/tools/blob/2aac38fc50ef0778c934859296f400f2dab764d8/claat/parser/md/parse.go#L443

Reply:
Actually, I found the usage below in `func button` but because this `isButton` always returns false, the condition never holds.

Reply:
I should have grepped for TODO… (but with the tests issue and no time given to work on those, this is why I didn't dare proposing it beforehand). Indeed, we don't recognize button, we don't use it on tutorials.ubuntu.com. I should spend time looking at it again, and what syntax would generate a button (if I remember correctly on gdoc, it was a link + special name).

I can't handle this today, but will probably be available to look into it later this week, is that fine?

Reply:
yeah, we can add it later. maybe @samtstern can help as well.